Writing Flow: Collapse range in horizontal edge check by selection direction

aduth opened the pull request:
Fixes #5095

This pull request seeks to resolve an issue where it is not possible to unexpand a selection from the beginning of a paragraph block. I'm not entirely clear the intent with collapsing range in the `isHorizontalEdge` function, though it would seem that it should be based on the direction of the current selection itself, not necessarily the direction being considered as the edge.

__Testing instructions:__

Verify that various shift-selection within and across blocks works as expected, particularly the previously failing behavior:

1. Insert a new paragraph block with some text
2. Make a selection of the text from the beginning of the block
3. Press Shift+ArrowLeft to unexpand the selection
4. Note that the selection becomes unexpanded as expected

aduth committed:
Writing Flow: Collapse range in horizontal edge check by direction

iseulde commented:
I'll have a look in a bit. Some previous art, but unsure why it is done that way: https://github.com/tinymce/tinymce/blob/d941efc13979b175e6c4c61946b631edc79b75bc/src/core/main/ts/api/dom/Selection.ts#L472-L491.

iseulde commented:
Hm, I'm not able to get this to work... even when pressing Shift+ArrowLeft furiously. :) The selection just seems stuck which I guess should collapse on this branch. I'll have a closer look at the code.

iseulde commented on utils/dom.js:
```diff
@@ -45,7 +75,9 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
-		range.collapse( isReverse );
+
+		// Collapse to the extent of the selection by direction.
+		range.collapse( ! isSelectionForward( selection ) );
```
Is the goal to set this range to the document? This is just used for calculation purposes (a clone) right?

iseulde replied:
I'm not quite sure here how the direction of a selection relates to this check. This would collapse a range at the edge of a field to the end or the start depending on direction (so in one case this will collapse to the end and the check fails), while we should check whether or not the range touches the edge? See https://github.com/WordPress/gutenberg/commit/1bb965830820a23ec5ab6768a432c01315a7e832

ellatrix committed:
Try alternative approach

iseulde commented:
@aduth Tried something different here: https://github.com/WordPress/gutenberg/commit/1bb965830820a23ec5ab6768a432c01315a7e832

Benefit is that it requires less code and no cloned range.

aduth commented:
@iseulde I like it, makes sense that we'd not care to consider the edge as breached if pressing arrows while having a non-callapsed selection (since the intent of this is to simply collapse in the direction of the arrow press).

Two thoughts though:

- Should this be the responsibility of `isHorizontalEdge` ? The name `isHorizontalEdge` doesn't imply that we'd be considering it strictly as in the context of an arrow press. Alternatively we could make this a combined condition `isSelectionCollapsed && isHorizontalEdge`, though not sure if there's a worrysome impact of calling `getSelection` or `getRange` twice.
- If we do keep it as-is, we should update the documentation for the `requireCollapsedRange` parameter

aduth commented:
Feel free to merge your commit into my branch as well.

aduth commented:
I'm going to write an end-to-end test for verifying this expected behavior.

iseulde commented:
> though not sure if there's a worrysome impact of calling getSelection or getRange twice.

This should be negligible and I'm not opposed to separating that. We'd have to generalise it for all input fields.

aduth commented:
A few more notes after playing with it a bit:

- Should we want to update `isVerticalEdge` as well?
- I think we want the inverse to be the default?

```diff
diff --git a/editor/components/writing-flow/index.js b/editor/components/writing-flow/index.js
index 48f7c08ce..25a0ba852 100644
--- a/editor/components/writing-flow/index.js
+++ b/editor/components/writing-flow/index.js
@@ -205,13 +205,13 @@ class WritingFlow extends Component {
 			// Moving from block multi-selection to single block selection
 			event.preventDefault();
 			this.moveSelection( isReverse );
-		} else if ( isVertical && isVerticalEdge( target, isReverse, isShift ) ) {
+		} else if ( isVertical && isVerticalEdge( target, isReverse, ! isShift ) ) {
 			const closestTabbable = this.getClosestTabbable( target, isReverse );
 			if ( closestTabbable ) {
 				placeCaretAtVerticalEdge( closestTabbable, isReverse, this.verticalRect );
 				event.preventDefault();
 			}
-		} else if ( isHorizontal && isHorizontalEdge( target, isReverse, isShift ) ) {
+		} else if ( isHorizontal && isHorizontalEdge( target, isReverse, ! isShift ) ) {
 			const closestTabbable = this.getClosestTabbable( target, isReverse );
 			placeCaretAtHorizontalEdge( closestTabbable, isReverse );
 			event.preventDefault();
diff --git a/utils/dom.js b/utils/dom.js
index 9e6c3f606..87ca3f18d 100644
--- a/utils/dom.js
+++ b/utils/dom.js
@@ -19,7 +19,7 @@ const { TEXT_NODE, ELEMENT_NODE } = window.Node;
  *
  * @return {boolean} True if at the horizontal edge, false if not.
  */
-export function isHorizontalEdge( container, isReverse, requireCollapsedRange = false ) {
+export function isHorizontalEdge( container, isReverse, requireCollapsedRange = true ) {
 	if ( includes( [ 'INPUT', 'TEXTAREA' ], container.tagName ) ) {
 		if ( container.selectionStart !== container.selectionEnd ) {
 			return false;
@@ -48,7 +48,7 @@ export function isHorizontalEdge( container, isReverse, requireCollapsedRange =
 		return false;
 	}
 
-	if ( ! requireCollapsedRange && ! range.collapsed ) {
+	if ( requireCollapsedRange && ! range.collapsed ) {
 		return false;
 	}
```

aduth commented:
Specifically, another behavior I'm noticing (on master and here prior to the above changes) is that when having a selection which extends to the beginning or end of a block, pressing the arrow should still just collapse the selection, not advance the caret to the previous / next block (at least this is how it behaves in Google Docs).

aduth committed:
Writing Flow: Move collapsed selection condition to WritingFlow

iseulde commented on utils/dom.js:
```diff
@@ -10,16 +10,30 @@ import tinymce from 'tinymce';
+ *
+ * @return {boolean} Whether there is a collapsed selection.
+ */
+export function hasCollapsedSelection() {
```
Will this work on `textarea` and `input`? Should a `container` be provided just like `isHorizontalEdge` etc.?

iseulde replied:
That's a good point, and we never actually seem to check if the selection is within the `container` in case of an `contenteditable` field. I do think we should provide the container (needed at least of `textarea` and `input`) and ensure _that_ contains a collapsed selection, not just the document in general? Or do you view this differently?

iseulde replied:
E.g. you might argue that since this is triggered by a keydown event, we can assume the selection is in the `container`.

aduth replied:
>not just the document in general

My thinking was toward the document selection in general. Like `container = document.activeElement`.

aduth commented:
Playing with the native textarea a bit, I think directionality has no impact here.

If there is an uncollapsed selection, pressing left or right should simply collapse to the extent in the direction of the arrow press.

If there is an uncollapsed selection, pressing up or down should move upward/downward as if the selection was collapsed.

aduth commented:
>If there is an uncollapsed selection, pressing up or down should move upward/downward as if the selection was collapsed.

Noting that with recent changes, this isn't working as expected. Thinking implementing the end-to-end tests sooner than later can help in iterating toward the ideal behavior.

iseulde commented on utils/dom.js:
```diff
@@ -10,16 +10,30 @@ import tinymce from 'tinymce';
+ * @return {boolean} Whether there is a collapsed selection.
+ */
+export function hasCollapsedSelection() {
+	const selection = window.getSelection();
```
There's also `selection.isCollapsed` which might work too. https://developer.mozilla.org/en-US/docs/Web/API/Selection

aduth replied:
>There's also `selection.isCollapsed` which might work too.

At that point, we may as well not even have this function, since `window.getSelection().collapsed` is barely more verbose.

iseulde replied:
Yeah :)

```js
const { getSelection } = window;

getSelection().isCollapsed;
```

iseulde commented on utils/dom.js:
```diff
@@ -10,16 +10,30 @@ import tinymce from 'tinymce';
+}
+
 /**
  * Check whether the caret is horizontally at the edge of the container.
```
Might want to update "caret" to "selection", as "caret" implies a collapsed selection I think.

aduth committed:
Writing Flow: Vertical navigate from uncollapsed selections

aduth committed:
Writing Flow: Use Selection#isCollapsed in place of hasCollapsedSelection

aduth committed:
Rich Text: Verify collapsed selection in delete behavior

aduth committed:
Utils: Improve range rect detection for br-separated content

aduth committed:
Testing: Add Writing Flow E2E tests

ellatrix committed:
Try getRectangleFromRange fix without manipulating live DOM

aduth commented on utils/dom.js:
```diff
@@ -150,14 +136,28 @@ export function getRectangleFromRange( range ) {
+	if ( startContainer.nodeType === ELEMENT_NODE ) {
+		const { childNodes } = startContainer;
+		const selectedNode = childNodes[
+			// Make sure there can be no errors.
```
Were you able to determine why it's the case that sometimes `startOffset` exceeds the number of `childNodes` ?

iseulde replied:
I've not encountered it. This is just to make sure it never does, if it would for some strange reason.

aduth replied:
Hmm, not entirely comfortable leaving this as an unknown.

iseulde replied:
Hm, me neither. Let's remove this check and address any problems if they come up?

aduth replied:
Yeah, I think so. Will give it another round of testing this morning.

aduth replied:
The issue with removing the safety check can be seen when, with two paragraphs, you press ArrowLeft from the second paragraph (to the first).

Specifically, it can occur that `startOffset` is `1` even though the first paragraph only has a single child node (where `startContainer.childNodes[ 1 ]` is `undefined`). Based on the [documentation for `startOffset`](https://developer.mozilla.org/en-US/docs/Web/API/Range/startOffset), this might be expected ("number of child nodes between [...]"), where perhaps the caret is not considered as being within the text node, but rather after it?

While setting a bound with `Math.min` avoids the error, I don't think it's returning an accurate result. In the above scenario, it returns a bounding rect for the entire text node, effectively the same box as the entire paragraph, whereas with a collapsed selection at the end of the paragraph, we'd expect it to be zero width at an X coordinate aligned to the end of the paragraph.

I'm still finding 573269eccfdccfb85ddb8e2421bef7a82e6f1a22 to be the most durable approach, despite the fact that it requires manipulating the DOM to retrieve the rect. I wonder if we could exhaust all reasonable attempts to retrieve the rect before using the temporary `span` technique only if we still cannot get a valid rect (where invalid is either `undefined` or one with `0`'d property values).

Annoyingly, this appears to be a main point of browser inconsistency / bugginess. For example, in [this snippet](http://jsbin.com/lavixus/1/edit?html,js,output), adding a new line to the contenteditable causes the rect to become zero'd in latest Chrome, despite a handful of resolved Chromium bug reports which have claimed to have addressed these sorts of issues ([[1]](https://bugs.chromium.org/p/chromium/issues/detail?id=435438), [[2]](https://bugs.chromium.org/p/chromium/issues/detail?id=574363)).

Seeing modules like [rangefix](https://github.com/edg2s/rangefix) has me sad for the state of browser compatibility on these functions, and sadder that our end-to-end approach won't be able to test against a broad set of browsers. Nothing I've found seems to work perfectly, aside from [`bounding-client-rect`](https://github.com/webmodules/bounding-client-rect/blob/master/index.js) which uses basically the same approach as in 573269eccfdccfb85ddb8e2421bef7a82e6f1a22 . If it's a point of frequent regression across non-Chromium browsers that our E2E suite can't cover well, we could always extract these out these normalization functions into a separate project which uses a more thorough end-to-end approach (e.g. [Selenium](https://www.seleniumhq.org/)).

aduth replied:
TinyMCE obviously handles this well, where replacing this line:

https://github.com/WordPress/gutenberg/blob/049645b01766f679ac245d1674371b7f5aa3028c/blocks/rich-text/index.js#L662

...with:

```js
rect = this.editor.selection.getBoundingClientRect();
```

...will return the value we expect. Unsure if we can/want to be able to leverage this directly from our generic DOM utilities here.

iseulde replied:
Yeah, this is what TinyMCE does:

https://github.com/tinymce/tinymce/blob/5f1e352a539359b351573d0621693e4580aa81d1/src/core/main/ts/caret/CaretPosition.ts#L126-L224

Maybe we can use this through `tinyMCE.dom.Selection`.

iseulde replied:
We're also using TinyMCE's isEmpty check here.

aduth committed:
Utils: Update caret to selection in edge tests comments

aduth committed:
Writing Flow: Invert position when isReverse not aligned to backward

iseulde commented on utils/dom.js:
```diff
@@ -36,38 +65,40 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
+	// selection direction. If `isReverse` is does not align to selection
+	// being backward, invert the position.
+	let position = isReverse ? 'start' : 'end';
+	if ( isReverse === isSelectionForward( selection ) ) {
```
Starting to wonder if we could do something like:

```js
function getDirectedRangeObject() {
	return {
		anchorNode: startContainer,
		anchorOffset: startOffset,
		focusNode: endContainer,
		focusOffset: endOffset,
	} = window.getSelection();

	return {
		startContainer,
		startOffset,
		endContainer,
		endOffset,
	};
}
```

iseulde replied:
Or even replace `start` and `end` with `anchor` and `focus`.

iseulde replied:
(And just look at the selection object, not care about ranges inside.)

iseulde commented on utils/dom.js:
```diff
@@ -36,38 +65,40 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
 		return false;
 	}
 
+	const order = isReverse ? 'first' : 'last';
```
Would this also need to invert? I haven't checked.

aduth replied:
>Would this also need to invert? I haven't checked.

I don't think it should need to, since this is for traversing through parents at the extent of the node, where selection direction doesn't matter.

ellatrix committed:
Try direct selection

ellatrix committed:
Only look at focusNode

aduth committed:
Utils: Document horizontal edge focusOffset

aduth commented:
Using focus offset appears to work well. Cherry-picked those commits into this branch.

aduth committed:
Utils: Use temporary text node for rect fallback

aduth committed:
Writing Flow: Test horizontal navigation as handled

Specifically for TinyMCE's overrides on inline boundary traversal

aduth commented:
I became a bit overzealous with end-to-end tests here and came up with an (admittedly not the most elegant but workable) solution for #5095.

aduth commented on utils/dom.js:
```diff
@@ -100,16 +130,8 @@ export function isVerticalEdge( container, isReverse, collapseRanges = false ) {
-	}
-
-	if ( ! range || ! range.collapsed ) {
+	const range = selection.rangeCount ? selection.getRangeAt( 0 ) : null;
```
Which logic are you referring to? Since `isVerticalEdge` is working primarily off the `rect`, I don't think we'll need similar logic with `focusNode` and such.

iseulde commented on utils/dom.js:
```diff
@@ -307,7 +284,7 @@ export function placeCaretAtVerticalEdge( container, isReverse, rect, mayUseScro
 	const buffer = rect.height / 2;
 	const editableRect = container.getBoundingClientRect();
-	const x = rect.left + ( rect.width / 2 );
+	const x = rect.left;
```
Just curious: what prompted this change?

aduth replied:
When pressing up from an uncollapsed selection, we want the caret to move directly above where the start of the selection occurs. Previously, it would place it above the half-way point of the selection.

![image](https://user-images.githubusercontent.com/1779930/39643417-b52a0792-4fa2-11e8-95f0-d081f0c41993.png)

iseulde commented on utils/dom.js:
```diff
@@ -150,21 +129,19 @@ export function getRectangleFromRange( range ) {
+	// `getClientRects` can be empty in some browsers. This can be resolved
+	// by adding a temporary text node to the range.
+	if ( ! rect ) {
+		const padNode = document.createTextNode( '\u200b' );
```
If this works for all cases in your testing, then cool, let's do it like this and I'll ponder there could be any better solution.

ellatrix reviewed (approved):
Very nice to have some e2e tests here!

aduth committed:
Element: Fix Fragment render error on empty children

aduth committed:
Testing: Add global guard against ZWSP in E2E content retrieval

aduth committed:
RichText: Remove ZWSP on focusout

aduth committed:
RichText: Consider emptiness by Children count utility

Since may be passed as fragment, not always as array.

aduth committed:
Writing Flow: Consider horizontal handled by stopPropagation in RichText

aduth committed:
RichText: Get content by converting tree node to element

format: 'tree' is only available in versions newer than 4.7.2, thus requiring update

aduth committed:
Writing Flow: Consider as edge by effective caret position

aduth commented:
I've spent the better part of the last day and a half hacking away at these behaviors, finding myself stumbling into several rabbit holes along the way. I'd been tempted to merge in its previous approved state, as it was largely functional, though I've noticed a few intermittent test failures which have prompted me to commit my local changes.

Some of them are only tangentially related to the goal, and given the number of them, it may be wise to fragment a few into separate pull requests against which to rebase. Or, we could bite the bullet and merge it as one.

Specifically, I noted several issues with our handling of inline boundaries, of TinyMCE's bogus nodes, and of lingering zero-width space characters.

A few of which can be reproduced on current master:

**Unable to horizontally traverse in a paragraph containing only inline boundary node:**

1. Add paragraph
2. Add a new paragraph _before_ the one created in Step 1
3. Cmd+B to bold
4. Write something
5. Press ArrowRight repeatedly
6. Get stuck at edge

Issue: TinyMCE leaves a bogus `br` node†, and our current logic to check horizontal edge doesn't account for this. In fact, our current test for [`tinymce.DOM.isEmpty`](https://github.com/WordPress/gutenberg/blob/65aaac80f64f4a73afa4b7acf1fedefe3a546ade/utils/dom.js#L40) is symptomatic of this larger problem, where in a contenteditable it's not possible for the caret to be traversed to certain nodes (empty nodes, `br`).

Resolution: We should check if the intended caret navigation will land in a reachable node. After a few iterations here, I finally landed on one similar to the original proposal, except not considering nodes and instead the effective caret position within text.

† It's my understanding this could be a bug within TinyMCE itself. In [TinyMCE Fiddle](http://fiddle.tinymce.com), usually the bogus `br` only exists so long as the field is empty.

Demo: https://codepen.io/aduth/pen/qYxPEM

**Held horizontal navigation from within inline boundary forces selection restore:**

1. Add paragraph
2. Add a new paragraph _before_ the one created in Step 1
3. Cmd+B to bold
4. Write something
5. Press ArrowRight, ArrowLeft, ArrowRight (the fact you need to do Left,Right is another bug, one in TinyMCE I think)
6. Type something (should be unbolded)
7. Save and reload
8. Place caret at start of first paragraph, hold ArrowRight
9. Notice the selection starts looping from the beginning of the selection

Issue: I haven't nailed this one down precisely, but I believe it has something to do with TinyMCE's selection restoration.

**Zero-width space character from inline boundaries saved to post content:**

1. Add paragraph
2. Cmd+B to bold
3. Write something
4. Press Enter to create a new paragraph
5. Press ArrowLeft
6. Save the post
7. The post is saved with a zero-width space character before the closing `</strong>` tag

Issue: We don't filter empty nodes in the [`getContent` implementation](https://github.com/WordPress/gutenberg/blob/65aaac80f64f4a73afa4b7acf1fedefe3a546ade/editor/components/rich-text/index.js#L721), so these are saved. The [`filterEmptyNodes` function](https://github.com/WordPress/gutenberg/blob/65aaac80f64f4a73afa4b7acf1fedefe3a546ade/editor/components/rich-text/index.js#L90) is an option here, but it occurs shallowly. Since TinyMCE performs this already in its own (HTML) serialization, it's unfortunately we'd need to recreate this. In newer versions we can use `{ format: 'tree' }` to receive a tree node which is cleaned of empty nodes. This works okay, with a few caveats:

- We still need to filter empty nodes in our `splitContent` function, and there's not an obvious way to leverage TinyMCE's internal cleanup logic for this, so we may still have need for the reimplemented node filtering.
- Since we're not converting from DOM to React and using `dom-react`'s attribute mapping, attributes aren't guaranteed to be in the canonical form for React. Fortunately our new serializer (#5897) doesn't care about this and will add them correctly. But as long as it's rendered in the editor, warnings could occur (keys and incorrect props properties).

**TinyMCE's inline boundary zero-width spaces linger after blurring the field, causing inconsistent edge traversal behavior:**

1. Add paragraph
2. Add a new paragraph
3. Cmd+B to bold
4. Write something
5. Press up to return to first paragraph
6. Press ArrowRight until you're in second paragraph
7. Note that the caret is _within_ the inline boundary
8. Press ArrowLeft to return to first paragraph
9. Press ArrowRight
10. Note that the caret is _before_ the inline boundary

Issue: TinyMCE's inline boundary zero-width character delineater remains when blurring the field, and the variation of its presence has an impact on where our `placeCaretAtHorizontalEdge` results in user behavior.

Resolution: Not implemented here, but there is some ["clean-up" that TinyMCE performs on `NodeChange`](https://github.com/tinymce/tinymce/blob/741db353b80fb6f164732994d7e50587feb7f400/src/core/main/ts/keyboard/BoundarySelection.ts#L104-L110). Naively adding this to the RichText's `focusout` had other regressing behavior. Will need to investigate. Ideally this would occur automatically on `focusout` by default in TinyMCE. 

**Inline boundary zero-width space occurs at beginning of node for new inline node:**

1. Add paragraph
2. Cmd+B to bold
3. Write something
4. In Developer Tools Console, enter `window.getSelection().focusNode`
5. Note something like `"&#65279;Bold"` (ZWSP at beginning)
6. Blur field
7. Click at end of the paragraph
8. In Developer Tools Console, enter `window.getSelection().focusNode`
9. Note something like `"&#65279;Bold"` (ZWSP at end, where caret is)

Issue: This is mainly problematic in detecting where the caret will land on a horizontal arrow press. It would be most convenient to check in a forward movement that the next character is a ZWSP, but in the above edge case this is not correct.

Resolution: A workaround has been implemented to check for ZWSP at the beginning for forward arrow movement.

aduth committed:
Rich Text: Return element fragment as array

May cause React to complain about lack of keys, but existing blocks map from value with expectation it's array (they probably shouldn't).

aduth committed:
Testing: Attempt to resolve race conditions with TinyMCE initialization

aduth commented:
cc @spocke as well re: confirming potential bugs noted above.

I've unfortunately needed to introduce a few "wait for" promises in the end-to-end tests, as it appears there can be some race conditions with TinyMCE readiness, or at least this appears to be the case.

iseulde commented:
> Issue: TinyMCE leaves a bogus br node†, and our current logic to check horizontal edge doesn't account for this. In fact, our current test for tinymce.DOM.isEmpty is symptomatic of this larger problem, where in a contenteditable it's not possible for the caret to be traversed to certain nodes (empty nodes, br).

The same issue would appear if there is a trailing BR node without the bogus attribute, so wouldn't it be better to ignore trailing BR nodes in general?

I'd prefer if we merge the previously approved state because I'm a bit confused with all the changes. Would be great if there's (a) PR(s) exactly fixing these cases where we can discuss the best solution and make sure there is no further performance degradation.

aduth commented:
>The same issue would appear if there is a trailing BR node without the bogus attribute, so wouldn't it be better to ignore trailing BR nodes in general?

Yes, the current implementation does this.

I've started fragmenting this pull request out into a few separate ones: #6710, #6711, #6712

aduth commented:
I've been looking at this again between Friday and today, and in addition to the [aforementioned pull requests](https://github.com/WordPress/gutenberg/pull/6467#issuecomment-388439283), there's a few separate tasks which could be tackled independently:

- Ensuring ZWSP are removed from text nodes
- Removing collapse consideration from `isHorizontalEdge`
- Caret vertical movement on uncollapsed selection

The ZWSP handling in particular has kept me thinking. While it would be nice to take advantage of TinyMCE's tree form, since we can't use it consistently (for filtering split content), I think we'll still want to have support for removing empty nodes from a DOM tree. I think there's an opportunity to consolidate many like behaviors:

- Cleaning ZWSP after focus leaves node
  - Use instead of `removeZwsp` as implemented in this pull request
- Removing bogus nodes from element value
  - Use instead of `filterEmptyNodes`
  - Use instead of `createTinyMCEElement`
  - Use instead of `this.editor.dom.isEmpty` check in `getContent` (return static reference after filtering, if filtered value is empty)
  - Use instead of `domToString` aggregating of fragment

Work-in-progress looks something like:

```js
export function domToElement( node ) {
	return nodeListToReact( node.childNodes );
}

export function domToString( node, editor ) {
	return editor ? editor.serializer.serialize( node ) : node.innerHTML;
}

export function removeEmptyChildNodes( node ) {
	for ( let i = 0; i < node.childNodes.length; i++ ) {
		const child = node.childNodes[ i ];

		switch ( child.nodeType ) {
			case Node.TEXT_NODE:
				if ( REGEXP_TINYMCE_ZWSP.test( child.nodeValue ) ) {
					child.nodeValue = child.nodeValue.replace( REGEXP_TINYMCE_ZWSP, '' );

					if ( ! child.nodeValue ) {
						node.removeChild( child );
					}
				}
				break;

			case Node.ELEMENT_NODE:
				if ( child.hasAttribute( 'data-mce-bogus' ) ) {
					if ( child.getAttribute( 'data-mce-bogus' ) !== 'all' ) {
						while ( child.firstChild ) {
							node.insertBefore( child.firstChild, child );
						}
					}

					node.removeChild( child );
				}
				break;
		}

		if ( child.nodeType === Node.TEXT_NODE && REGEXP_TINYMCE_ZWSP.test( child.nodeValue ) ) {
			child.nodeValue = child.nodeValue.replace( REGEXP_TINYMCE_ZWSP, '' );

			if ( ! child.nodeValue ) {
				node.removeChild( child );
			}
		}

		removeEmptyChildNodes( child );
	}
}

export function domToFormat( nodes, format, editor ) {
	const fragment = document.createDocumentFragment();
	for ( let i = 0; i < nodes.length; i++ ) {
		fragment.appendChild( nodes[ i ].cloneNode( true ) );
	}

	removeEmptyChildNodes( fragment );

	switch ( format ) {
		case 'string':
			return domToString( fragment, editor );
		default:
			return domToElement( fragment );
	}
}
```

Needs to be particularly conscious of performance impact of creating / cloning nodes. Not sure if copying node attributes to an object form might be faster than `cloneNode`.

There's more "clean-up" which needs to occur to strip TinyMCE attributes and placeholders in the filtering function. Ideally we could leverage some of TinyMCE's internals to avoid recreating this, or at least it should be based on a like implementation.

aduth commented:
This has stagnated and is beyond saving at this point. Many of the ideas have inspired other pull requests, and yet more ideas included here are still valid and could be pulled out into their own separate pull requests, which I'll explore.
